app/assets/src/components/views/discovery/DiscoverySidebar.jsx

```diff
@@ -0,0 +1,213 @@
+import React from "React";
+import cx from "classnames";
+import {
+  uniqBy,
+  sum,
+  sumBy,
+  flatten,
+  map,
+  keyBy,
+  countBy,
+  sortBy
+} from "lodash/fp";
+import moment from "moment";
+
+import PropTypes from "~/components/utils/propTypes";
+import { getAllHostGenomes } from "~/api";
+
+import cs from "./discovery_sidebar.scss";
+
+export default class DiscoverySidebar extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      stats: {
+        samples: 0,
+        projects: 0
+        // TODO (gdingle):
+        // avg_reads_per_sample: 0,
+      },
+      metadata: {
+        host: {},
+        tissue: {},
+        created_at: {}
+        // TODO (gdingle):
+        // location: {},
+      },
+      _computed: null
+    };
+  }
+
+  async componentDidMount() {
+    this.genomes = await getAllHostGenomes();
+  }
+
+  componentDidUpdate() {
+    const { currentTab, projects } = this.props;
+
+    // TODO (gdingle): this is not working when flipping back from visualization tab
+    if (this.state._computed && this.state._computed == currentTab) {
+      // Only update once because of
```

Comment:
@MarkAZhang Any ideas on best practice for avoiding this issue?

Reply:
Yea, I wanted to ask in person someone because the update behavior is beyond me. `componentDidUpdate` is firing thousands of times.

Reply:
I've encountered this issue before -- in my case it was when I had inadvertently *called* a function in a prop, instead of just *passing* it as the prop. So that's why I thought maybe a similar hard-to-spot mistake is at work here. Hope Mark knows more.

Reply:
When you call `setState`, it causes the component to re-render again, which causes `componentDidUpdate` to fire again. If you call `setState` again in `componentDidUpdate`, this will cause the loop to happen again, creating an infinite loop.

A common approach is to use the `prevProps` or `prevState` parameters that are passed into `componentDidUpdate` to conditionally call `setState` only when the relevant thing has changed.

An example from `BareDropdown`:

```
  componentDidUpdate(prevProps) {
    if (
      prevProps.items !== this.props.items ||
      prevProps.options !== this.props.options ||
      prevProps.value !== this.props.value
    ) {
      this.setState({
        filteredItems: this.getFilteredItems(this.state.filterString)
      });
    }
  }
```

In your case, I think you can do `prevProps.currentTab !== this.props.currentTab`

Reply:
See https://reactjs.org/docs/react-component.html#componentdidupdate for more info.

Reply:
Thanks for the explanation. It’s been too long since I used react! I found the clear solution is to use getDerivedState method.